tpmutil/encoding.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2018, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tpmutil
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+// LengthPrefixSize is the size in bytes of length prefix for byte slices.
+//
+// In TPM 1.2 this is 4 bytes.
+// In TPM 2.0 this is 2 bytes.
+var LengthPrefixSize int
+
+// packedSize computes the size of a sequence of types that can be passed to
+// binary.Read or binary.Write.
+func packedSize(elts ...interface{}) (int, error) {
+	var size int
+	for _, e := range elts {
+		v := reflect.ValueOf(e)
+		switch v.Kind() {
+		case reflect.Ptr:
+			s, err := packedSize(reflect.Indirect(v).Interface())
+			if err != nil {
+				return 0, err
+			}
+
+			size += s
+		case reflect.Struct:
+			for i := 0; i < v.NumField(); i++ {
+				s, err := packedSize(v.Field(i).Interface())
+				if err != nil {
+					return 0, err
+				}
+
+				size += s
+			}
+		case reflect.Slice:
+			switch s := e.(type) {
+			case []byte:
+				size += LengthPrefixSize + len(s)
+			case RawBytes:
+				size += len(s)
+			default:
+				return 0, fmt.Errorf("encoding of %T is not supported, only []byte and RawBytes slices are", e)
+			}
+		default:
+			s := binary.Size(e)
+			if s < 0 {
+				return 0, fmt.Errorf("can't calculate size of type %T", e)
+			}
+
+			size += s
+		}
+	}
+
+	return size, nil
+}
+
+// packWithHeader takes a header and a sequence of elements that are either of
+// fixed length or slices of fixed-length types and packs them into a single
+// byte array using binary.Write. It updates the CommandHeader to have the right
+// length.
+func packWithHeader(ch commandHeader, cmd ...interface{}) ([]byte, error) {
+	hdrSize := binary.Size(ch)
+	bodySize, err := packedSize(cmd...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't compute packed size for message body: %v", err)
+	}
+	ch.Size = uint32(hdrSize + bodySize)
+	in := []interface{}{ch}
+	in = append(in, cmd...)
+	return Pack(in...)
+}
+
+// Pack encodes a set of elements into a single byte array, using
+// encoding/binary. This means that all the elements must be encodeable
+// according to the rules of encoding/binary.
+//
+// It has one difference from encoding/binary: it encodes byte slices with a
+// prepended uint16 length, to match how the TPM encodes variable-length
```

Comment:
I think this comment has gone stale (and probably other comments). Instead of a `uint16` for the size, this now depends on `LengthPrefixSize`, and this comment should say that.

Reply:
Removed length type from comment. LengthPrefixSize is not exported now.
Added package comment about UseTPMXXLengthPrefixSize functions instead.